### ISSUE TITLE
Switch from TSLint to typescript-eslint

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [JoshuaKGoldberg/TypeStat#335]
This repository still uses TSLint, which is being deprecated: https://github.com/palantir/tslint/issues/4534. It'd be great to move it over to [typescript-eslint](http://typescript-eslint.io) to no longer use deprecated packages.

Protip: [tslint-to-eslint-config](https://github.com/typescript-eslint/tslint-to-eslint-config) is the recommended package to do so!

**Comment by arvidboivie** [comment]
I'm having a go at this!

**Comment by arvidboivie** [comment]
I'm having issues running `verify` on the project, even without any changes. I'm probably missing something easy, and don't want to go making changes to get this working. The tests are failing, here's the output:

```
 FAIL  src/shared/NameGenerator.test.ts
  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    /Users/arbo/repos/hacktoberfest/TypeStat/src/shared/NameGenerator.ts:12
      countsPerBase = new Map();
                    ^

    SyntaxError: Unexpected token =

    > 1 | import { NameGenerator } from "./NameGenerator";
        | ^
      2 |
      3 | describe("NameGenerator", () => {
      4 |     it("creates a basic name when the base portion hasn't been requested yet", () => {

      at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:537:17)
      at ScriptTransformer.transform (node_modules/@jest/transform/build/ScriptTransformer.js:579:25)
      at Object.<anonymous> (src/shared/NameGenerator.test.ts:1:1)

 FAIL  src/cli/runCli.test.ts
  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    /Users/arbo/repos/hacktoberfest/TypeStat/src/shared/FileInfoCache.ts:17
      nodeReferences = new Map();
                     ^

    SyntaxError: Unexpected token =

      4 | import { TypeStatOptions } from "../../options/types";
      5 | import { LazyCache } from "../../services/LazyCache";
    > 6 | import { FileInfoCache } from "../../shared/FileInfoCache";
        | ^
      7 | import { convertMapToObject, Dictionary } from "../../shared/maps";
      8 | import { NameGenerator } from "../../shared/NameGenerator";
      9 | import { collectFilteredNodes } from "../collectFilteredNodes";

      at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:537:17)
      at ScriptTransformer.transform (node_modules/@jest/transform/build/ScriptTransformer.js:579:25)
      at Object.<anonymous> (src/runtime/providers/createCoreMutationsProvider.ts:6:1)
```

**Comment by JoshuaKGoldberg** [comment]
Oh 😦 I don't get those errors locally on a fresh clone @arvidboivie, working on Windows. Maybe the issue is Mac/*nix specific?

Anyway, you can ignore them - you shouldn't need to run `yarn run verify` for this issue. Just running `yarn run eslint` after it's created should be sufficient. Does that help?

**Comment by arvidboivie** [comment]
@JoshuaKGoldberg You're right, I skipped looking at those errors. Probably Mac related. Tried my hand at replacing the rules, PR up now.

**Comment by JoshuaKGoldberg** [comment]
Fixed by #371.
